sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -830,6 +831,21 @@ object PushProjectionThroughUnion extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Pushes Project operator to Limit operator.
+ */
+object PushProjectionThroughLimit extends Rule[LogicalPlan] {
```

Comment:
Why don't you fix `SpecialLimits` to add `CollectLimitExec` in these cases instead of switching it?

Reply:
Let me think it over.

Reply:
I have solved the above scenario by adding a new match case in SpecialLimits.